File: Benchmarking/Python3/matinv.py
def read_matrix(filename):
    with open(filename, 'r') as f:
        # Read the size of the matrix
        n = int(f.readline())
        
        # Read the matrix elements
        matrix = [[float(elem) for elem in f.readline().split()] for _ in range(n)]
    
    return matrix

def write_matrix_to_file(filename, matrix):
    with open(filename, 'w') as f:
        for row in matrix:
            f.write(" ".join(f"{elem}" for elem in row) + "\n")

def gauss_jordan_inverse(matrix):
    # Get matrix length
    n = len(matrix)
    
    # Create an augmented matrix with the identity matrix on the right
    augmented_matrix = [row + [1 if i == j else 0 for j in range(n)] for i, row in enumerate(matrix)]
    
    # Perform Gauss-Jordan elimination
    for i in range(n):
        # Find the pivot element
        pivot = augmented_matrix[i][i]
        if pivot == 0:
            # Swap with a row below that has a non-zero pivot
            for j in range(i + 1, n):
                if augmented_matrix[j][i] != 0:
                    augmented_matrix[i], augmented_matrix[j] = augmented_matrix[j], augmented_matrix[i]
                    pivot = augmented_matrix[i][i]
                    break
            else:
                raise ValueError("Matrix is singular and cannot be inverted.")
        
        # Normalize the pivot row
        for j in range(2 * n):
            augmented_matrix[i][j] /= pivot
        
        # Eliminate the current column in all other rows
        for j in range(n):
            if j != i:
                factor = augmented_matrix[j][i]
                for k in range(2 * n):
                    augmented_matrix[j][k] -= factor * augmented_matrix[i][k]
    
    # Extract the inverse matrix from the augmented matrix
    inverse_matrix = [row[n:] for row in augmented_matrix]
    
    return inverse_matrix

# Main program
matrix = read_matrix("../random_matrix.txt")
inverse_matrix = gauss_jordan_inverse(matrix)

write_matrix_to_file("./inverse_matrix.txt", inverse_matrix)

File: Benchmarking/Pascal/matinv.pas
program MatrixInverse;

type
  TMatrix = array of array of Real;

var
  matrix, inverse: TMatrix;
  n: Integer;

procedure ReadMatrix(const filename: string; var m: TMatrix);
var
  f: Text;
  i, j: Integer;
begin
  Assign(f, filename);
  Reset(f);

  ReadLn(f, n);
  SetLength(m, n, n);

  for i := 0 to n - 1 do
    for j := 0 to n - 1 do
      Read(f, m[i][j]);

  Close(f);
end;

procedure WriteMatrixToFile(const filename: string; const m: TMatrix);
var
  f: Text;
  i, j: Integer;
begin
  Assign(f, filename);
  Rewrite(f);

  for i := 0 to Length(m) - 1 do
  begin
    for j := 0 to Length(m[i]) - 1 do
      Write(f, m[i][j]:0:6, ' ');  // Adjusted to 6 decimal places
    WriteLn(f);
  end;

  Close(f);
end;

procedure GaussJordanInverse(const matrix: TMatrix; var inverse: TMatrix);
var
  i, j, k: Integer;
  pivot, factor: Real;
  augmented: TMatrix;
begin
  n := Length(matrix);
  SetLength(augmented, n, 2 * n);

  // Create an augmented matrix with the identity matrix on the right
  for i := 0 to n - 1 do
  begin
    for j := 0 to n - 1 do
    begin
      augmented[i][j] := matrix[i][j];
      if i = j then
        augmented[i][j + n] := 1.0
      else
        augmented[i][j + n] := 0.0;
    end;
  end;

  // Perform Gauss-Jordan elimination
  for i := 0 to n - 1 do
  begin
    pivot := augmented[i][i];
    if pivot = 0 then
    begin
      // Find a row to swap with
      k := i + 1;
      while (k < n) and (augmented[k][i] = 0) do
        Inc(k);

      if k = n then
      begin
        WriteLn('Matrix is singular and cannot be inverted.');
        Halt(1);
      end;

      // Swap rows
      for j := 0 to 2 * n - 1 do
      begin
        factor := augmented[i][j];
        augmented[i][j] := augmented[k][j];
        augmented[k][j] := factor;
      end;

      pivot := augmented[i][i];
    end;

    // Normalize the pivot row
    for j := 0 to 2 * n - 1 do
      augmented[i][j] := augmented[i][j] / pivot;

    // Eliminate the current column in all other rows
    for k := 0 to n - 1 do
    begin
      if k <> i then
      begin
        factor := augmented[k][i];
        for j := 0 to 2 * n - 1 do
          augmented[k][j] := augmented[k][j] - factor * augmented[i][j];
      end;
    end;
  end;

  // Extract the inverse matrix from the augmented matrix
  SetLength(inverse, n, n);
  for i := 0 to n - 1 do
    for j := 0 to n - 1 do
      inverse[i][j] := augmented[i][j + n];
end;

begin
  // Read matrix elements from file
  ReadMatrix('../random_matrix.txt', matrix);

  // Compute the inverse matrix
  SetLength(inverse, Length(matrix), Length(matrix[0]));
  GaussJordanInverse(matrix, inverse);

  // Write the inverse matrix to a file
  WriteMatrixToFile('./inverse_matrix.txt', inverse);
end.
